From 9c27581f0fe28a313f10b71c782cc99ad045c67b Mon Sep 17 00:00:00 2001
From: Sergey Suloev <ssuloev@orpaltech.com>
Date: Sat, 4 Jan 2020 20:16:07 +0300
Subject: [PATCH] input: Use reset controller in edt-ft5x06 driver

---
 drivers/input/touchscreen/edt-ft5x06.c | 27 +++++++++++---------------
 1 file changed, 11 insertions(+), 16 deletions(-)

diff --git a/drivers/input/touchscreen/edt-ft5x06.c b/drivers/input/touchscreen/edt-ft5x06.c
index d61731c..305bad8 100644
--- a/drivers/input/touchscreen/edt-ft5x06.c
+++ b/drivers/input/touchscreen/edt-ft5x06.c
@@ -27,8 +27,9 @@
 #include <linux/gpio/consumer.h>
 #include <linux/input/mt.h>
 #include <linux/input/touchscreen.h>
-#include <asm/unaligned.h>
 #include <linux/regulator/consumer.h>
+#include <linux/reset.h>
+#include <asm/unaligned.h>
 
 #define WORK_REGISTER_THRESHOLD		0x00
 #define WORK_REGISTER_REPORT_RATE	0x08
@@ -91,8 +92,8 @@ struct edt_ft5x06_ts_data {
 	u16 num_y;
 	struct regulator *vcc;
 
-	struct gpio_desc *reset_gpio;
 	struct gpio_desc *wake_gpio;
+	struct reset_control *reset;
 
 #if defined(CONFIG_DEBUG_FS)
 	struct dentry *debug_dir;
@@ -1094,12 +1095,10 @@ static int edt_ft5x06_ts_probe(struct i2c_client *client,
 	if (error)
 		return error;
 
-	tsdata->reset_gpio = devm_gpiod_get_optional(&client->dev,
-						     "reset", GPIOD_OUT_HIGH);
-	if (IS_ERR(tsdata->reset_gpio)) {
-		error = PTR_ERR(tsdata->reset_gpio);
-		dev_err(&client->dev,
-			"Failed to request GPIO reset pin, error %d\n", error);
+	tsdata->reset = devm_reset_control_get(&client->dev, "reset");
+	if (IS_ERR(tsdata->reset)) {
+		error = PTR_ERR(tsdata->reset);
+		dev_err(&client->dev, "couldn't get reset line: %d\n", error);
 		return error;
 	}
 
@@ -1117,11 +1116,8 @@ static int edt_ft5x06_ts_probe(struct i2c_client *client,
 		gpiod_set_value_cansleep(tsdata->wake_gpio, 1);
 	}
 
-	if (tsdata->reset_gpio) {
-		usleep_range(5000, 6000);
-		gpiod_set_value_cansleep(tsdata->reset_gpio, 0);
-		msleep(300);
-	}
+	/* reset the panel*/
+	reset_control_reset(tsdata->reset);
 
 	input = devm_input_allocate_device(&client->dev);
 	if (!input) {
@@ -1203,10 +1199,9 @@ static int edt_ft5x06_ts_probe(struct i2c_client *client,
 	device_init_wakeup(&client->dev, 1);
 
 	dev_dbg(&client->dev,
-		"EDT FT5x06 initialized: IRQ %d, WAKE pin %d, Reset pin %d.\n",
+		"EDT FT5x06 initialized: IRQ %d, WAKE pin %d.\n",
 		client->irq,
-		tsdata->wake_gpio ? desc_to_gpio(tsdata->wake_gpio) : -1,
-		tsdata->reset_gpio ? desc_to_gpio(tsdata->reset_gpio) : -1);
+		tsdata->wake_gpio ? desc_to_gpio(tsdata->wake_gpio) : -1);
 
 	return 0;
 }
-- 
2.17.1

