From 93bbabe3c56fec2a664fa6de6d81b38794129378 Mon Sep 17 00:00:00 2001
From: Sergey Suloev <ssuloev@orpaltech.com>
Date: Sun, 13 May 2018 22:56:01 +0300
Subject: [PATCH] rpi: Maximum kernel image size to 64MB

---
 include/configs/rpi.h | 3 +++
 1 file changed, 3 insertions(+)

diff --git a/include/configs/rpi.h b/include/configs/rpi.h
index 69a22e1..406b1f5 100644
--- a/include/configs/rpi.h
+++ b/include/configs/rpi.h
@@ -23,6 +23,9 @@
 	(&((struct bcm2835_timer_regs *)BCM2835_TIMER_PHYSADDR)->clo)
 #endif
 
+/* Increase the maximum kernel image size from the default to 64 MB */
+#define CONFIG_SYS_BOOTM_LEN	(64 * 1024 * 1024)
+
 /*
  * 2835 is a SKU in a series for which the 2708 is the first or primary SoC,
  * so 2708 has historically been used rather than a dedicated 2835 ID.
-- 
2.17.0

