From 02f7404cc6323b3dd65207e82220bce481a633ef Mon Sep 17 00:00:00 2001
From: Sergey Suloev <ssuloev@orpaltech.com>
Date: Thu, 2 Aug 2018 21:05:23 +0300
Subject: [PATCH] arm: enable DT-overlay support

---
 arch/arm/Kconfig | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index 00b2848..dffb94b 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -533,6 +533,8 @@ config ARCH_BCM283X
 	select OF_CONTROL
 	select PL01X_SERIAL
 	select SERIAL_SEARCH_ALL
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	imply FAT_WRITE
 
 config TARGET_VEXPRESS_CA15_TC2
@@ -795,10 +797,11 @@ config ARCH_SUNXI
 	select USB_STORAGE if DISTRO_DEFAULTS
 	select USB_KEYBOARD if DISTRO_DEFAULTS
 	select USE_TINY_PRINTF
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	imply CMD_GPT
 	imply DISTRO_DEFAULTS
 	imply FAT_WRITE
-	imply OF_LIBFDT_OVERLAY
 	imply PRE_CONSOLE_BUFFER
 	imply SPL_GPIO_SUPPORT
 	imply SPL_LIBCOMMON_SUPPORT
-- 
2.18.0

