From 6db3bf017bfdba7be68e0104ecd63aa308c6a228 Mon Sep 17 00:00:00 2001
From: Sergey Suloev <ssuloev@orpaltech.com>
Date: Wed, 7 Feb 2018 13:58:54 +0300
Subject: [PATCH] Add more logs

---
 cmd/fdt.c | 7 +++++++
 1 file changed, 7 insertions(+)

diff --git a/cmd/fdt.c b/cmd/fdt.c
index b783b0d..5e1d530 100644
--- a/cmd/fdt.c
+++ b/cmd/fdt.c
@@ -235,6 +235,8 @@ static int do_fdt(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
 
 		nodeoffset = fdt_path_offset (working_fdt, pathp);
 		if (nodeoffset < 0) {
+			printf("make a new node: pathp=%s, nodep=%s\n", pathp, nodep);
+
 			/*
 			 * Not found or something else bad happened.
 			 */
@@ -272,6 +274,8 @@ static int do_fdt(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
 
 		nodeoffset = fdt_path_offset (working_fdt, pathp);
 		if (nodeoffset < 0) {
+			printf("set a prop value: pathp=%s, prop=%s\n", pathp, prop);
+
 			/*
 			 * Not found or something else bad happened.
 			 */
@@ -332,6 +336,8 @@ static int do_fdt(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
 
 		nodeoffset = fdt_path_offset(working_fdt, pathp);
 		if (nodeoffset < 0) {
+			printf("get a prop value: pathp=%s, prop=%s\n", pathp, prop);
+
 			/*
 			 * Not found or something else bad happened.
 			 */
@@ -461,6 +467,7 @@ static int do_fdt(cmd_tbl_t *cmdtp, int flag, int argc, char * const argv[])
 		 */
 		nodeoffset = fdt_path_offset (working_fdt, argv[2]);
 		if (nodeoffset < 0) {
+			printf("remove a prop/node: pathp=%s\n", argv[2]);
 			/*
 			 * Not found or something else bad happened.
 			 */
-- 
2.16.1

