From e2205b1bddb28e84dc473e423a461f4d2a1b1ba9 Mon Sep 17 00:00:00 2001
From: Sergey Suloev <ssuloev@orpaltech.com>
Date: Fri, 31 Jan 2020 01:19:34 +0300
Subject: [PATCH] arm: enable DT-overlay support

---
 arch/arm/Kconfig | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index 36c9c2f..a4d750c 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -613,6 +613,8 @@ config ARCH_BCM283X
 	select OF_CONTROL
 	select PL01X_SERIAL
 	select SERIAL_SEARCH_ALL
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	imply CMD_DM
 	imply FAT_WRITE
 
@@ -972,13 +974,14 @@ config ARCH_SUNXI
 	select USB_KEYBOARD if DISTRO_DEFAULTS
 	select USB_STORAGE if DISTRO_DEFAULTS
 	select SPL_USE_TINY_PRINTF
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	imply CMD_DM
 	imply CMD_GPT
 	imply CMD_UBI if MTD_RAW_NAND
 	imply DISTRO_DEFAULTS
 	imply FAT_WRITE
 	imply FIT
-	imply OF_LIBFDT_OVERLAY
 	imply PRE_CONSOLE_BUFFER
 	imply SPL_GPIO_SUPPORT
 	imply SPL_LIBCOMMON_SUPPORT
-- 
2.17.1

