From 69e8d1e3a325894a07519473168313dd91221e35 Mon Sep 17 00:00:00 2001
From: ssuloev <ssuloev@orpaltech.com>
Date: Thu, 21 Apr 2022 14:24:29 +0300
Subject: [PATCH] arm: enable DT-overlay support

---
 arch/arm/Kconfig | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index 4567c18..91a74a2 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -606,6 +606,8 @@ config ARCH_BCM283X
 	select DM_SERIAL
 	select GPIO_EXTRA_HEADER
 	select OF_CONTROL
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	select PL01X_SERIAL
 	select SERIAL_SEARCH_ALL
 	imply CMD_DM
@@ -1072,6 +1074,8 @@ config ARCH_SUNXI
 	select OF_BOARD_SETUP
 	select OF_CONTROL
 	select OF_SEPARATE
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	select SPECIFY_CONSOLE_INDEX
 	select SPL_SEPARATE_BSS if SPL
 	select SPL_STACK_R if SPL
@@ -1093,7 +1097,6 @@ config ARCH_SUNXI
 	imply DISTRO_DEFAULTS
 	imply FAT_WRITE
 	imply FIT
-	imply OF_LIBFDT_OVERLAY
 	imply PRE_CONSOLE_BUFFER
 	imply SPL_GPIO
 	imply SPL_LIBCOMMON_SUPPORT
-- 
2.25.1

