From 9678d3432edf47a1830c10c9160c731e3326fbb6 Mon Sep 17 00:00:00 2001
From: ssuloev <ssuloev@orpaltech.com>
Date: Fri, 5 May 2023 17:08:01 +0300
Subject: [PATCH] arm: enable DT-overlay support

---
 arch/arm/Kconfig | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index 8a1e223..f2ea49c 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -629,6 +629,8 @@ config ARCH_BCM283X
 	select DM_SERIAL
 	select GPIO_EXTRA_HEADER
 	select OF_CONTROL
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	select PL01X_SERIAL
 	select SERIAL_SEARCH_ALL
 	imply CMD_DM
@@ -1110,6 +1112,8 @@ config ARCH_SUNXI
 	select OF_BOARD_SETUP
 	select OF_CONTROL
 	select OF_SEPARATE
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	select PINCTRL
 	select SPECIFY_CONSOLE_INDEX
 	select SPL_SEPARATE_BSS if SPL
@@ -1132,7 +1136,6 @@ config ARCH_SUNXI
 	imply DISTRO_DEFAULTS
 	imply FAT_WRITE
 	imply FIT
-	imply OF_LIBFDT_OVERLAY
 	imply PRE_CONSOLE_BUFFER
 	imply SPL_GPIO
 	imply SPL_LIBCOMMON_SUPPORT
-- 
2.25.1

