From d5a134836a02576e735ec1604e968eb7e5be9064 Mon Sep 17 00:00:00 2001
From: Sergey Suloev <sergey.suloev@gmail.com>
Date: Thu, 21 Aug 2025 22:07:57 +0300
Subject: [PATCH] arm: Enable DT-overlay support

---
 arch/arm/Kconfig | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index 6ff3f27..a17ca04 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -673,6 +673,8 @@ config ARCH_BCM283X
 	select DM_SERIAL
 	select GPIO_EXTRA_HEADER
 	select OF_CONTROL
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	select PL01X_SERIAL
 	select SERIAL_SEARCH_ALL
 	imply CMD_DM
@@ -1194,6 +1196,8 @@ config ARCH_SUNXI
 	select MMU_PGPROT if ARM64
 	select OF_BOARD_SETUP
 	select OF_CONTROL
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	select PINCTRL
 	select SPECIFY_CONSOLE_INDEX
 	select SPL_SEPARATE_BSS if SPL
@@ -1217,7 +1221,6 @@ config ARCH_SUNXI
 	imply DM_REGULATOR_FIXED
 	imply FAT_WRITE
 	imply FIT
-	imply OF_LIBFDT_OVERLAY
 	imply PRE_CONSOLE_BUFFER
 	imply SPL_GPIO
 	imply SPL_LIBCOMMON_SUPPORT
-- 
2.43.0

