From 61a96cdeaa6de0868cf63e47bfa69ced19841dfe Mon Sep 17 00:00:00 2001
From: Sergey Suloev <ssuloev@orpaltech.com>
Date: Tue, 15 Feb 2022 18:49:53 +0300
Subject: [PATCH] clk/bcm/rpi: Do not exceed end of firmware clocks

---
 drivers/clk/bcm/clk-raspberrypi.c | 5 +++--
 1 file changed, 3 insertions(+), 2 deletions(-)

diff --git a/drivers/clk/bcm/clk-raspberrypi.c b/drivers/clk/bcm/clk-raspberrypi.c
index dd3b71e..29f9da1 100644
--- a/drivers/clk/bcm/clk-raspberrypi.c
+++ b/drivers/clk/bcm/clk-raspberrypi.c
@@ -248,7 +248,7 @@ struct rpi_firmware_get_clocks_response {
 static int raspberrypi_discover_clocks(struct raspberrypi_clk *rpi,
 				       struct clk_hw_onecell_data *data)
 {
-	struct rpi_firmware_get_clocks_response *clks;
+	struct rpi_firmware_get_clocks_response *clks, *clks_end;
 	int ret;
 
 	clks = devm_kcalloc(rpi->dev,
@@ -263,7 +263,8 @@ static int raspberrypi_discover_clocks(struct raspberrypi_clk *rpi,
 	if (ret)
 		return ret;
 
-	while (clks->id) {
+	clks_end = clks + RPI_FIRMWARE_NUM_CLK_ID;
+	while (clks < clks_end) {
 		struct clk_hw *hw;
 
 		switch (clks->id) {
-- 
2.25.1

