From fefaf2b875639c918025d1446073954724a462e2 Mon Sep 17 00:00:00 2001
From: Sergey Suloev <ssuloev@orpaltech.com>
Date: Fri, 18 Jan 2019 02:15:00 +0300
Subject: [PATCH] input: Use reset controller in edt-ft5x06 driver

---
 drivers/input/touchscreen/edt-ft5x06.c | 52 ++++++++++++++++----------
 1 file changed, 33 insertions(+), 19 deletions(-)

diff --git a/drivers/input/touchscreen/edt-ft5x06.c b/drivers/input/touchscreen/edt-ft5x06.c
index 1e18ca0..bd3156a 100644
--- a/drivers/input/touchscreen/edt-ft5x06.c
+++ b/drivers/input/touchscreen/edt-ft5x06.c
@@ -25,6 +25,7 @@
  *    http://www.glyn.com/Products/Displays
  */
 
+#define DEBUG
 #include <linux/module.h>
 #include <linux/ratelimit.h>
 #include <linux/irq.h>
@@ -39,6 +40,8 @@
 #include <linux/input/mt.h>
 #include <linux/input/touchscreen.h>
 #include <linux/of_device.h>
+#include <linux/regulator/consumer.h>
+#include <linux/reset.h>
 
 #define WORK_REGISTER_THRESHOLD		0x00
 #define WORK_REGISTER_REPORT_RATE	0x08
@@ -86,17 +89,18 @@ struct edt_reg_addr {
 };
 
 struct edt_ft5x06_ts_data {
-	struct i2c_client *client;
-	struct input_dev *input;
+	struct i2c_client	*client;
+	struct input_dev	*input;
 	struct touchscreen_properties prop;
 	u16 num_x;
 	u16 num_y;
 
-	struct gpio_desc *reset_gpio;
-	struct gpio_desc *wake_gpio;
+	struct gpio_desc	*wake_gpio;
+	struct regulator	*power;
+	struct reset_control	*rstc;
 
 #if defined(CONFIG_DEBUG_FS)
-	struct dentry *debug_dir;
+	struct dentry	*debug_dir;
 	u8 *raw_buffer;
 	size_t raw_bufsize;
 #endif
@@ -991,12 +995,19 @@ static int edt_ft5x06_ts_probe(struct i2c_client *client,
 
 	tsdata->max_support_points = chip_data->max_support_points;
 
-	tsdata->reset_gpio = devm_gpiod_get_optional(&client->dev,
-						     "reset", GPIOD_OUT_HIGH);
-	if (IS_ERR(tsdata->reset_gpio)) {
-		error = PTR_ERR(tsdata->reset_gpio);
+	tsdata->power = devm_regulator_get(&client->dev, "power");
+	if (IS_ERR(tsdata->power)) {
+		error = PTR_ERR(tsdata->power);
+		dev_err(&client->dev,
+			"couldn't get our power regulator: %d\n", error);
+		return error;
+	}
+
+	tsdata->rstc = devm_reset_control_get_shared(&client->dev, "reset");
+	if (IS_ERR(tsdata->rstc)) {
+		error = PTR_ERR(tsdata->rstc);
 		dev_err(&client->dev,
-			"Failed to request GPIO reset pin, error %d\n", error);
+			"couldn't get our reset line: %d\n", error);
 		return error;
 	}
 
@@ -1005,7 +1016,7 @@ static int edt_ft5x06_ts_probe(struct i2c_client *client,
 	if (IS_ERR(tsdata->wake_gpio)) {
 		error = PTR_ERR(tsdata->wake_gpio);
 		dev_err(&client->dev,
-			"Failed to request GPIO wake pin, error %d\n", error);
+			"failed to request GPIO wake pin: %d\n", error);
 		return error;
 	}
 
@@ -1014,11 +1025,15 @@ static int edt_ft5x06_ts_probe(struct i2c_client *client,
 		gpiod_set_value_cansleep(tsdata->wake_gpio, 1);
 	}
 
-	if (tsdata->reset_gpio) {
-		usleep_range(5000, 6000);
-		gpiod_set_value_cansleep(tsdata->reset_gpio, 0);
-		msleep(300);
-	}
+	/* power up */
+	error = regulator_enable(tsdata->power);
+	if (error)
+		return error;
+
+	/* reset the panel*/
+	msleep(20);
+	reset_control_reset(tsdata->rstc);
+	msleep(300);
 
 	input = devm_input_allocate_device(&client->dev);
 	if (!input) {
@@ -1100,10 +1115,9 @@ static int edt_ft5x06_ts_probe(struct i2c_client *client,
 	device_init_wakeup(&client->dev, 1);
 
 	dev_dbg(&client->dev,
-		"EDT FT5x06 initialized: IRQ %d, WAKE pin %d, Reset pin %d.\n",
+		"EDT FT5x06 initialized: IRQ %d, WAKE pin %d.\n",
 		client->irq,
-		tsdata->wake_gpio ? desc_to_gpio(tsdata->wake_gpio) : -1,
-		tsdata->reset_gpio ? desc_to_gpio(tsdata->reset_gpio) : -1);
+		tsdata->wake_gpio ? desc_to_gpio(tsdata->wake_gpio) : -1);
 
 	return 0;
 }
-- 
2.17.1

