From 5d1979b0b2331eb2774294fb949a2b6c77b32969 Mon Sep 17 00:00:00 2001
From: Sergey Suloev <ssuloev@orpaltech.com>
Date: Fri, 2 Feb 2018 00:28:13 +0300
Subject: [PATCH] Enable DT-overlay -support

---
 arch/arm/Kconfig | 4 ++++
 1 file changed, 4 insertions(+)

diff --git a/arch/arm/Kconfig b/arch/arm/Kconfig
index f2c35e3..d48463f 100644
--- a/arch/arm/Kconfig
+++ b/arch/arm/Kconfig
@@ -500,6 +500,8 @@ config ARCH_BCM283X
 	select DM_SERIAL
 	select DM_GPIO
 	select OF_CONTROL
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	imply FAT_WRITE
 
 config TARGET_VEXPRESS_CA15_TC2
@@ -713,6 +715,8 @@ config ARCH_SUNXI
 	select USB_STORAGE if DISTRO_DEFAULTS
 	select USB_KEYBOARD if DISTRO_DEFAULTS
 	select USE_TINY_PRINTF
+	select OF_LIBFDT
+	select OF_LIBFDT_OVERLAY
 	imply CMD_GPT
 	imply FAT_WRITE
 	imply PRE_CONSOLE_BUFFER
-- 
2.16.1

