From 1f18e2a068d2ab06ec3f845a9a54d1841c17e2be Mon Sep 17 00:00:00 2001
From: ssuloev <ssuloev@orpaltech.com>
Date: Thu, 3 Oct 2024 22:05:01 +0300
Subject: [PATCH] of/property/fw_devlink: Support allwinner,sram links

---
 drivers/of/property.c | 22 ++++++++++++++++++++++
 1 file changed, 22 insertions(+)

diff --git a/drivers/of/property.c b/drivers/of/property.c
index 11b922f..dc9aa96 100644
--- a/drivers/of/property.c
+++ b/drivers/of/property.c
@@ -1351,6 +1351,27 @@ static struct device_node *parse_remote_endpoint(struct device_node *np,
 	return of_graph_get_remote_port_parent(np);
 }
 
+static struct device_node *parse_allwinner_sram(struct device_node *np,
+						const char *prop_name, int index)
+{
+	struct device_node *sram_node;
+
+	if (!IS_ENABLED(CONFIG_SUNXI_SRAM))
+		return NULL;
+
+	if (strcmp(prop_name, "allwinner,sram"))
+		return NULL;
+
+	if (index > 0)
+		return NULL;
+
+	sram_node = of_parse_phandle(np, prop_name, 0);
+	sram_node = of_get_parent(sram_node);
+	sram_node = of_get_parent(sram_node);
+
+	return sram_node;
+}
+
 static const struct supplier_bindings of_supplier_bindings[] = {
 	{ .parse_prop = parse_clocks, },
 	{ .parse_prop = parse_interconnects, },
@@ -1399,6 +1420,7 @@ static const struct supplier_bindings of_supplier_bindings[] = {
 		.parse_prop = parse_post_init_providers,
 		.fwlink_flags = FWLINK_FLAG_IGNORE,
 	},
+	{ .parse_prop = parse_allwinner_sram, },
 	{}
 };
 
-- 
2.34.1

