From 48af82760043dd8108ca4273b722f99b3110272b Mon Sep 17 00:00:00 2001
From: Sergey Suloev <sergey.suloev@gmail.com>
Date: Mon, 28 Jul 2025 17:09:37 +0300
Subject: [PATCH] spi/sun4i: Fix SPI-mode on resume in sun4i SPI-driver

---
 drivers/spi/spi-sun4i.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/spi/spi-sun4i.c b/drivers/spi/spi-sun4i.c
index aa92fd5..c12afbe 100644
--- a/drivers/spi/spi-sun4i.c
+++ b/drivers/spi/spi-sun4i.c
@@ -393,6 +393,7 @@ static int sun4i_spi_runtime_resume(struct device *dev)
 	struct spi_controller *host = dev_get_drvdata(dev);
 	struct sun4i_spi *sspi = spi_controller_get_devdata(host);
 	int ret;
+	u32 reg;
 
 	ret = clk_prepare_enable(sspi->hclk);
 	if (ret) {
@@ -406,8 +407,10 @@ static int sun4i_spi_runtime_resume(struct device *dev)
 		goto err;
 	}
 
+	reg = sun4i_spi_read(sspi, SUN4I_CTL_REG);
+
 	sun4i_spi_write(sspi, SUN4I_CTL_REG,
-			SUN4I_CTL_MASTER | SUN4I_CTL_TP);
+			reg | SUN4I_CTL_MASTER | SUN4I_CTL_TP);
 
 	return 0;
 
-- 
2.43.0

