From c1b469511d5103e5c1913aabc2616104f81bd644 Mon Sep 17 00:00:00 2001
From: Sergey Suloev <ssuloev@orpaltech.com>
Date: Sat, 6 Jan 2018 01:12:14 +0300
Subject: [PATCH] Set max kernel size to 64MB

---
 include/configs/rpi.h | 2 ++
 1 file changed, 2 insertions(+)

diff --git a/include/configs/rpi.h b/include/configs/rpi.h
index c499b45..e3fa4ab 100644
--- a/include/configs/rpi.h
+++ b/include/configs/rpi.h
@@ -160,5 +160,7 @@
 	ENV_MEM_LAYOUT_SETTINGS \
 	BOOTENV
 
+/* Increase the maximum kernel image size from the default to 64 MB */
+#define CONFIG_SYS_BOOTM_LEN (64 * 1024 * 1024)
 
 #endif
-- 
2.15.1

